otter/convergence/composition.py

```diff
@@ -66,17 +68,23 @@ def json_to_LBConfigs(lbs_json):
     return lbd
 
 
-def get_desired_group_state(launch_config, desired):
+def get_desired_group_state(group_id, launch_config, desired,
+                            _prepare_lc=prepare_launch_config):
     """
     Create a :obj:`DesiredGroupState` from a launch config and desired
     number of servers.
 
+    :param str group_id: Scaling group ID
     :param dict launch_config: Group's launch config as per
         :obj:`otter.json_schema.group_schemas.launch_config`
     :param int desired: Group's desired capacity
+    :param callable _prepare_lc: Optional implementation of
+        `prepare_launch_config` useful for testing
     """
     lbs = json_to_LBConfigs(launch_config['args']['loadBalancers'])
+    server_lc = {'server': launch_config['args']['server']}
     desired_state = DesiredGroupState(
-        launch_config={'server': launch_config['args']['server']},
+        launch_configs=(_prepare_lc(group_id, server_lc)
+                        for _ in itertools.count()),
```

Comment:
This is really surprising. Why are we defining launch_configs as an infinite iterable of launch configs? I take it prepare_launch_config has side-effects (some random number stuff?) but this solution strikes me as very unexpected.

There's really no difference between this and just adding a function called "get_launch_config" that returns a new, randomized launch config. Using an iterable just means that code that wants to access this has to use `next` to get the return value, which is a pretty strange interface. I think it's more reasonable to call prepare_launch_config directly from that convergence code than to hide this side-effect in what used to be a simple data structure.